stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/usecase/MarkAllReadImpl.kt

```diff
@@ -0,0 +1,32 @@
+package io.getstream.chat.android.livedata.usecase
+
+import io.getstream.chat.android.client.call.Call
+import io.getstream.chat.android.client.call.CoroutineCall
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.livedata.ChatDomainImpl
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+
+public interface MarkAllRead {
+    public operator fun invoke(): Call<Boolean>
+}
+
+internal data class MarkAllReadImpl(private val domain: ChatDomainImpl) : MarkAllRead {
+    override fun invoke(): Call<Boolean> = CoroutineCall(domain.scope) {
+        domain.scope.let { scope ->
+            scope.async(Dispatchers.IO) {
+                // update the UI first
+                domain.allActiveChannels().map { channel ->
+                    scope.async(Dispatchers.Main) {
+                        channel.markRead()
+                    }
+                }.awaitAll() // wait for the UI updates to avoid races
+
+                // then update via remote API
+                domain.client.markAllRead().execute()
+                Result(true, null)
+            }
+        }.await()
+    }
```

Comment:
```suggestion
    override fun invoke(): Call<Boolean> {
        return CoroutineCall(domain.scope) {
            // update the UI first
            domain.allActiveChannels().map { channel ->
                domain.scope.async(Dispatchers.Main) {
                    channel.markRead()
                }
            }.awaitAll() // wait for the UI updates to avoid races

            // then update via remote API
            domain.client.markAllRead().await()
            Result(true, null)
        }
    }
```

Reply:
Is there a reason you prefer `return`?

Reply:
Just a heads up - if I use `domain.client.markAllRead().await()`, `MarkAllReadImplTest` hangs indefinitely. I've used `suspendCancellableCoroutine` to wrap callbacks in the past, so I'm not sure why it's not working in this instance. I'll check it out later if I get some time.